### ISSUE TITLE
Method inference issue with extension methods that have lamdas as argument when the class implementation contains a method overload with object argument.

### ISSUE THREAD

**Issue by nsarris** [dotnet/roslyn#57759]
**Version Used**:  c# 10.0

**Steps to Reproduce**:

1. Create a class with a method with an object argument.
```
public class Test
    {
        public void Do(object value)
        {

        }
}
```

2. Create an a generic extension method with an Expression<Func<T>> argument or a Func<T> argument.

```
public static class Extensions
    {
        public static void Do<T>(this Test test, Expression<Func<T>> expression)
        {

        }
    }
```

or

```
public static class Extensions
    {
        public static void Do<T>(this Test test, Func<T> expression)
        {

        }
    }
```

3. Call the method with a lambda arguemt 
```
new Test().Do(() => new DateTime());
```

**Expected Behavior**:

Expression argument should target the extension method with matching argument. 

**Actual Behavior**:

The class member with object overload is targeted. 

Please also note that behaviour is only manifested on extension methods.  Having the method as an overload in the class instead of an extension method results in the expected behaviour. If both overloads for Func<T> and Expression<Func<T>> are added then the extension methods are both ignored, while the instance methods fail with inference ambigiuity (which means they are infered properly). Removing the class implementation with object argument reverts to the expected behaviour. This behaviour is c# 10.0 only, previous versions displayed the expected behaviour.

 In short, there seems to be a bug with method inference wit extension methods when the arguments are lambdas when the cass implementation contains a method overload with object argument.

**Comment by CyrusNajmabadi** [comment]
@cston for final explanation, but I'm fairly certain this is by design and a breaking change we knowingly accepted. 

If an instance method is viable, we pick it before moving to extension methods. And in this case the insurance method is viable as a lambda is convertible to an object directly now in c#10

**Comment by cston** [comment]
Yes, the C#10 overload resolution behavior is expected.

The compiler can infer a delegate type of `Func<DateTime>` for the lambda expression above. As a result, with C#10 the lambda expression is considered implicitly convertible to `object` (and implicitly convertible to `MultcastDelegate` and `LambdaExpression`), so the instance method is applicable.

The breaking change is described in [Compiler Breaking Changes - DotNet 6.md](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%206.md) (see 2a), although the instance method in that document uses `Delegate` rather than `object`. I will add an `object` example as well.

**Comment by nsarris** [comment]
Ok it makes sense now, I was not aware of the breaking changes document, thank you, I will close the issue.
